crates/sui-indexer/src/indexer_reader.rs

```diff
@@ -1433,6 +1599,56 @@ impl IndexerReader {
     }
 }
 
+#[async_trait::async_trait]
+impl ObjectProvider for IndexerReader {
+    type Error = IndexerError;
+
+    async fn get_object(
```

Comment:
I couldn't see where this is called so it's hard to tell the implication. But I am a bit concerned about the performance  given that the interface does a single point read each time. Ideally during a GraphQL query we should try to group object queries as much as we can, instead of making individual queries.

Reply:
(note that this isn't used by GraphQL, where we use the DataLoader pattern to do the batching for us).

Reply:
it's not obvious indeed, I had a comment above
> mostly moved from `transactions.rs` other than object_changes parts, the move was necessary b/c get_object_changes needs an ObjectProvider 
https://github.com/MystenLabs/sui/blob/5505e624feccbd623f130358653e1fe613fd3181/crates/sui-json-rpc/src/object_changes.rs#L14

Reply:
this is how FN json rpc does it today, iiuc graphql does not use either the json rpc type or the util function here; it's possible that the util function can be improved, but prob not what this pr is about.

Reply:
yeah perf wise we may want to implement object changes differently...out of curiosity today we store balance change events in the db maybe we just keep storing object changes there as well till such time as we deprecate and remove jsonrpc?

Reply:
keeping object_changes is not end of the world while cleaning it up is still strictly better as
- more code sharing with FN means less change to diverge on migration from FN to indexer
- not significant but still cleaner schema and less storage
wdyt?

Reply:
Orthogonal to this pr, but I dug deeper and found the reason that graphql only needs `effects` is the definition differences of `object_changes` 
- graphql uses sui-type `ObjectChange` [definition](https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/effects/mod.rs#L378-L384), which is leaner and aligned with effects v2
- while JSON object change [definition](https://github.com/MystenLabs/sui/blob/main/crates/sui-json-rpc-types/src/object_changes.rs#L18) is enriched, especially for `Published` and `Transferred`, thus it needs extra parameters like ObjectProvider and sender to enrich.

maybe sth worth considering regarding the overall migration from JSON RPC to GraphQL cc @amnn

Reply:
The reason we can do this in GraphQL is that it's possible for people to request the enriched data through nested queries (i.e. we use the `sui-type` `ObjectChange`, but then if you want to know who an object was transferred to, or if the package that was created was a published package, you can write nested queries that extract the necessary information from the objects and fetching those objects is made efficient using batching), so there may be a noticeable (and not acceptable) regression in the schema if you do the same thing in JSON-RPC but don't proactively fetch those objects.